src/gui/qgsmaptoolcapture.cpp

```diff
@@ -1073,3 +1108,191 @@ void QgsMapToolCapture::updateExtraSnapLayer()
   }
 }
 
+
+
+void QgsMapToolCapture::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
```

Comment:
Why is this all moved to the base class? There's a lot here which isn't generic logic, e.g. handling the avoid intersections behaviour, force segmentisation if the layer doesn't support curves, etc. This logic will break existing subclasses, e.g. the annotation item digitization tools which don't have a vector layer associated with them.

Reply:
My idea is also that things like intersections could be optional on demand.

But having everything in the base class will benefit to other map tools (add part, add/fill ring).

Reply:
Hmm... I think this should be at least opt in only. But really, I don't think it has any place in a base QgsMapToolCapture class, which can be used to capture ANY geometry for any purpose, where all the avoid intersection/etc logic is unwanted. E.g. imagine a (very theoretical) "measure area of shape" tool which allows you to draw a rect/circle/... and shows the corresponding area in the status bar... there's no way you'd want geometries drawn using this tool to magically trigger the avoid intersections logic!

If you want to keep this stuff reusable, I'd insert a new class between QgsMapToolCapture and QgsMapToolDigitizeFeature, allowing the QgsMapToolCapture class to be kept for pure geometry captures only without any changes to the project/layer states.

Reply:
I have added an intermediate class `QgsMapToolCaptureLayerGeometry` in 47c98e0636f0eef3d22063ecc23396c84f72bccd